# Ex A: Node level statistics

## Q A.1

In [2]:
import numpy as np

A = np.array([
    [0., 0., 1., 1., 0., 1., 0.],
    [0., 0., 0., 0., 1., 1., 1.],
    [1., 0., 0., 1., 0., 1., 0.],
    [1., 0., 1., 0., 0., 1., 0.],
    [0., 1., 0., 0., 0., 1., 1.],
    [1., 1., 1., 1., 1., 0., 1.],
    [0., 1., 0., 0., 1., 1., 0.]
])

lambdas, E = np.linalg.eig(A)

print(np.round(lambdas, 3))

[ 3.646  2.    -1.646 -1.    -1.    -1.    -1.   ]


The largest eigenvector is the first one (3.646), thus we look at the first eigenvector.

In [3]:
print(np.round(E[:, 0], 3))

[-0.339 -0.339 -0.339 -0.339 -0.339 -0.558 -0.339]


Here we can see that node six (index 5) is the is the most central with an absolute value of 0.558.
This is being reinforced by looking at the adjacency matrix, where it has the most edges.

## Q A.2

In [4]:
from numpy.linalg import inv, matrix_power

size = A.shape[0]
I = np.eye(size)
D = np.diag(np.sum(A, axis=1))

# From lecture slides: c = (D(D-I))^-1 diag(A^3)
c = inv(D @ (D - I)) @ np.diag(matrix_power(A, 3))
print(c)

[1.  1.  1.  1.  1.  0.4 1. ]


# Ex B: Random walks

## Q B.1

The final distribution is the starting distribution p propagated through the random walk transition matrix inv(D)A.
So for each edge (1 in A) we divide by how many edges it has (D).

## Q B.2

For node one we have e1. If we take the adjacency matrix to the power of t, we get the t's node neighbors. If t=2 we get node one's neighbors' neighbors. Each t essentially gives another hop in the graph. Then we have that (A^t)e1. The totla number of distinct paths of length t, from node 1, is just the sum of the function (A^t)e1.

## Ex D: Programming exercise